src/webview/js/handleInboundEvents.js

```diff
@@ -0,0 +1,42 @@
+/* @flow strict-local */
+import invariant from 'invariant';
+import { ensureUnreachable } from '../../generics';
+
+import type { EditSequence } from '../generateInboundEventEditSequence';
+
+// TODO: Put more of the WebView's event-handling code in here.
+
+/**
+ * PRIVATE - don't use until we have more test coverage (coming soon)
+ */
+export function applyEditSequence(editSequence: EditSequence) {
+  const msglistElementsDiv = document.querySelector('div#msglist-elements');
+  invariant(msglistElementsDiv, 'applyEditSequence: expected msglistElementsDiv');
+
+  editSequence.forEach(edit => {
+    switch (edit.type) {
+      case 'insert': {
+        // TODO: Performance in the common case of many consecutive inserts:
+        //   https://github.com/zulip/zulip-mobile/pull/5188#discussion_r793030286
+        const newChild = document.createElement('div');
+        msglistElementsDiv.insertBefore(newChild, msglistElementsDiv.children[edit.index] ?? null);
+        newChild.outerHTML = edit.html;
+        break;
+      }
+      case 'replace': {
+        const newChild = document.createElement('div');
+        msglistElementsDiv.replaceChild(newChild, msglistElementsDiv.children[edit.index]);
+        newChild.outerHTML = edit.html;
+        break;
```

Comment:
Seems like the `.outerHTML =` should come before the `replaceChild` -- that way the element gets replaced atomically with the intended new version, without an intermediate state where it's been replaced by an empty div.

(Similarly I think the `insert` case would feel cleaner with the new child fully set up before it gets inserted, though for that one I don't see a concrete problem the premature insert causes other than perhaps performance.)

Reply:
Ah: if an element doesn't have a parent, then setting `.outerHTML` on it [won't work](https://w3c.github.io/DOM-Parsing/#dom-element-outerhtml).

> that way the element gets replaced atomically with the intended new version, without an intermediate state where it's been replaced by an empty div

> Similarly I think the `insert` case would feel cleaner with the new child fully set up before it gets inserted, though for that one I don't see a concrete problem the premature insert causes other than perhaps performance.

Mmm, I'm remembering: I think these concerns were what I had in mind for `DocumentFragment` to address in an earlier revision. (I should have written that down somewhere. 🙂) When you asked if we could remove the fragments, I guess I forgot, since I agreed that the fragments were pretty un-streamlined for performance.

Reply:
Hmmm, I see.

OK, DocumentFragment it is, then, like in the original revision. Chat discussion was here:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/dom.20manipulation/near/1319720